code/game/machinery/computer/arena.dm

```diff
@@ -0,0 +1,420 @@
+	for(var/team in teams)
+		var/datum/atom_hud/antag/teamhud = team_huds[team]
+		if(!teamhud) //These will be shared between arenas because this stuff is expensive and cross arena fighting is not a thing anyway
+			teamhud = new
+			teamhud.icon_color = team //change to lookup table
+			GLOB.huds += teamhud
+			team_huds[team] = teamhud
+			team_hud_index[team] = length(GLOB.huds)
+
+/**
+  * Loads the arenas from config directory.
+  * THESE ARE FULLY CACHED FOR QUICK SWITCHING SO KEEP TRACK OF THE AMOUNT
+  */
+/obj/machinery/computer/arena/proc/LoadDefaultArenas()
+	if(default_arenas_loaded)
+		return
+	var/arena_dir = "[global.config.directory]/arenas/"
+	var/list/default_arenas = flist(arena_dir)
+	for(var/arena_file in default_arenas)
+		var/simple_name = replacetext(replacetext(arena_file,arena_dir,""),".dmm","")
+		add_new_arena_template(null,arena_dir + arena_file,simple_name)
+
+/obj/machinery/computer/arena/proc/get_landmark_turf(landmark_tag)
+	for(var/obj/effect/landmark/arena/L in GLOB.landmarks_list)
+		if(L.arena_id == arena_id && L.landmark_tag == landmark_tag && isturf(L.loc))
+			return L.loc
+
+/obj/machinery/computer/arena/proc/get_load_point()
+	var/turf/A = get_landmark_turf(ARENA_CORNER_A)
+	var/turf/B = get_landmark_turf(ARENA_CORNER_B)
+	return locate(min(A.x,B.x),min(A.y,B.y),A.z)
+
+/obj/machinery/computer/arena/proc/get_arena_turfs()
+	var/lp = get_load_point()
+	var/turf/A = get_landmark_turf(ARENA_CORNER_A)
+	var/turf/B = get_landmark_turf(ARENA_CORNER_B)
+	var/turf/hp = locate(max(A.x,B.x),max(A.y,B.y),A.z)
+	return block(lp,hp)
+
+/obj/machinery/computer/arena/proc/clear_arena()
+	for(var/turf/T in get_arena_turfs())
+		T.empty(turf_type = /turf/open/indestructible)
+	current_arena_template = "None"
+
+/obj/machinery/computer/arena/proc/load_arena(arena_template,mob/user)
+	if(loading)
+		return
+	var/datum/map_template/M = arena_templates[arena_template]
+	if(!M)
+		to_chat(user,"<span class='warning'>No such arena</span>")
+		return
+	clear_arena() //Clear current arena
+	var/turf/A = get_landmark_turf(ARENA_CORNER_A)
+	var/turf/B = get_landmark_turf(ARENA_CORNER_B)
+	var/wh = abs(A.x - B.x)
+	var/hz = abs(A.y - B.y)
+	if(M.width > wh || M.height > hz)
+		to_chat(user,"<span class='warning'>Arena template is too big for the current arena!</span>")
+		return
+	loading = TRUE
+	var/bd = M.load(get_load_point())
+	if(bd)
+		current_arena_template = arena_template
+		arena_afterload(arena_template)
+	loading = FALSE
+	
+	message_admins("[key_name_admin(user)] loaded [arena_template] event arena for [arena_id] arena.")
+	log_admin("[key_name(user)] loaded [arena_template] event arena for [arena_id] arena.")
+
+	
+
+/// Handle arena specific special effects here.
+/obj/machinery/computer/arena/proc/arena_afterload(arena_filename)
+	/* Example code
+	switch(arena_filename)
+		if("example.dmm")
+			to_chat(world,"EXAMPLE ARENA FIGHT STARTING NOW. OR WHATEVER")
+		if("ripandtear.dmm")
+			give_everyone_chainsaws()
+	*/
```

Comment:
Wouldn't it be better to use map template datums for this? Add procs there that get run? bit better OOP wise.